File: app/Models/Product.php
<?php

namespace App\Models;

use App\Models\Traits\BelongsToCompany;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\HasOne;

class Product extends Model
{
    use HasFactory, BelongsToCompany;

    protected $fillable = [
        'company_id',
        'name',
        'sku',
        'barcode',
        'category_id',
        'description',
        'cost_price',
        'selling_price',
        'tax_rate',
        'image_path',
        'is_active',
    ];

    protected $casts = [
        'cost_price' => 'decimal:2',
        'selling_price' => 'decimal:2',
        'tax_rate' => 'decimal:2',
        'is_active' => 'boolean',
    ];

    public function category(): BelongsTo
    {
        return $this->belongsTo(Category::class);
    }

    public function inventory(): HasOne
    {
        return $this->hasOne(Inventory::class);
    }

    public function stockAdjustments(): HasMany
    {
        return $this->hasMany(StockAdjustment::class);
    }

    public function transactionItems(): HasMany
    {
        return $this->hasMany(TransactionItem::class);
    }

    public function getStockQuantityAttribute(): int
    {
        return $this->inventory?->quantity ?? 0;
    }

    public function isLowStock(): bool
    {
        if (!$this->inventory) {
            return true;
        }
        return $this->inventory->quantity <= $this->inventory->low_stock_threshold;
    }

    public function getPriceWithTaxAttribute(): float
    {
        return $this->selling_price * (1 + $this->tax_rate / 100);
    }

    public function scopeActive($query)
    {
        return $query->where('is_active', true);
    }

    public function scopeSearch($query, $term)
    {
        return $query->where(function ($q) use ($term) {
            $q->where('name', 'like', "%{$term}%")
              ->orWhere('sku', 'like', "%{$term}%")
              ->orWhere('barcode', 'like', "%{$term}%");
        });
    }
}

File: mobile_app/lib/shared/models/product.dart
class Product {
  final int id;
  final String name;
  final String? sku;
  final String? barcode;
  final double sellingPrice;
  final double taxRate;
  final int stock;
  final int lowStockThreshold;
  final bool isLowStock;
  final String? imageUrl;
  final Category? category;

  Product({
    required this.id,
    required this.name,
    this.sku,
    this.barcode,
    required this.sellingPrice,
    required this.taxRate,
    required this.stock,
    this.lowStockThreshold = 10,
    this.isLowStock = false,
    this.imageUrl,
    this.category,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'],
      name: json['name'],
      sku: json['sku'],
      barcode: json['barcode'],
      sellingPrice: (json['selling_price'] ?? 0).toDouble(),
      taxRate: (json['tax_rate'] ?? 0).toDouble(),
      stock: json['stock'] ?? 0,
      lowStockThreshold: json['low_stock_threshold'] ?? 10,
      isLowStock: json['is_low_stock'] ?? false,
      imageUrl: json['image_url'],
      category: json['category'] != null ? Category.fromJson(json['category']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'sku': sku,
      'barcode': barcode,
      'selling_price': sellingPrice,
      'tax_rate': taxRate,
      'stock': stock,
      'low_stock_threshold': lowStockThreshold,
      'is_low_stock': isLowStock,
      'image_url': imageUrl,
      'category': category?.toJson(),
    };
  }

  double get priceWithTax => sellingPrice * (1 + taxRate / 100);

  bool get isInStock => stock > 0;
}

class Category {
  final int id;
  final String name;
  final String? description;
  final int productsCount;
  final int? parentId;

  Category({
    required this.id,
    required this.name,
    this.description,
    this.productsCount = 0,
    this.parentId,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      productsCount: json['products_count'] ?? 0,
      parentId: json['parent_id'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'products_count': productsCount,
      'parent_id': parentId,
    };
  }
}
